doc/design.md

```diff
@@ -0,0 +1,922 @@
+In many cases the code is scattered around the project, in the implementation of individual operations.
+This is to be cleaned up when implementing the full set of operations.
+
+At a later point, a small extension of `Dimensions` can provide support for multi-dimensional chunking/partitioning.
+This could enable MPI support, cache blocking for more efficient use of the CPU caches, and working with data that does not fit into memory.
+- Store the full dimension extent in addition to the chunk extent in `Dimensions`.
+- If a dimension of a dataset represents just a subsection of the full dimension extent we use a special dimension tag.
+- Operations that operate on a particular dimension will then fail.
+  *Example: `Algorithms::DiffractionFocussing` requires `Dim::Spectrum`.*
+- Based on the stored full dimension extent we can assemble a dataset from chunks, or provide a multi-step operation that can deal with partial dimensions and does the assembling process.
+  *Example: A version of `Algorithms::DiffractionFocussing` that first focusses within each chunk and merges focussed data from all chunks in a final step.*
+
+
+### <a name="design-components-dimension-tags"></a>Dimension tags
+
+A dimension is identified by a tag.
+This is simply an `enum class Dim`.
+*Note:
+In the prototype this is currently named `enum class Dimension`, but we suggest renaming this to `enum class Dim` for brevity.
+Despite the abbreviation this should be clear and may actually improve code readability since it is more concise when used frequently?*
+
+An incomplete list of dimension tags:
+- `Dim::X`, `Dim::Y`, `Dim::Z`.
+- `Dim::Qx`, `Dim::Qy`, `Dim::Qz`.
+- `Dim::Detector`
+- `Dim::Spectrum`
+- `Dim::Tof`
+- `Dim::Wavelength`
+- `Dim::DetectorScan`
+- `Dim::Temperature`
+- `Dim::Polarization`
+- `Dim::Run`
+- `Dim::Row`
+- `Dim::Event`
+
+In many cases there is a corresponding coordinate tag for a dimension.
+
+
+### <a name="design-components-datasetview"></a>`class DatasetView`
+
+Due to the type-erasure we cannot simply treat `Dataset` as an array of structs, i.e., there is no equivalent to `API::MatrixWorkspace::getSpectrum(const size_t index)`.
+If only access to a single variable is required it can be access directly.
+For cases that require join access to multiple variables we provide `class DatasetView`:
+- Effectively provides a "zip-iterator", with support for automatic on-the-fly broadcast or transposition of dimensions.
+- Nested view allows for accessing a histogram-like view.
+- Allows for joint iteration with bin-edge variables using a `Bin` wrapper, overcoming the old nuisance of having `length+1` bin edges.
+
+The current implementation seems to support all required features.
+Performance was originally a problem due to the multi-dimensional iteration, but has been subsequently been optimized to the point where it should be sufficient for most applications.
```

Comment:
Is there any benchmark reference for performance?

Reply:
https://github.com/mantidproject/dataset/blob/master/benchmark/dataset_view_benchmark.cpp

Result on my workstation (memory bandwidth peaks is roughly 40-50 GB/s, I believe):

```
BM_DatasetView_mixed_dimension_addition/8/real_time                                  13724 ns      13724 ns      51842   13.0289GB/s   555.899M items/s
BM_DatasetView_mixed_dimension_addition/16/real_time                                 24137 ns      24137 ns      29091   14.8167GB/s   632.177M items/s
BM_DatasetView_mixed_dimension_addition/32/real_time                                 46396 ns      46396 ns      15030   15.4162GB/s   657.758M items/s
BM_DatasetView_mixed_dimension_addition/64/real_time                                103235 ns     103235 ns       6807   13.8568GB/s   591.225M items/s
BM_DatasetView_mixed_dimension_addition/128/real_time                               204474 ns     204470 ns       3636   13.9921GB/s   596.996M items/s
BM_DatasetView_mixed_dimension_addition/256/real_time                               380310 ns     380309 ns       2010   15.0457GB/s   641.951M items/s
BM_DatasetView_mixed_dimension_addition/512/real_time                               770289 ns     770277 ns        837   14.8569GB/s   633.894M items/s
BM_DatasetView_mixed_dimension_addition/1024/real_time                             1575921 ns    1575903 ns        475   14.5237GB/s   619.677M items/s
BM_DatasetView_mixed_dimension_addition/2048/real_time                             3093578 ns    3093570 ns        225   14.7972GB/s   631.348M items/s
BM_DatasetView_mixed_dimension_addition/4096/real_time                             5638200 ns    5638175 ns        119   16.2379GB/s   692.819M items/s
BM_DatasetView_mixed_dimension_addition/8192/real_time                            12360720 ns   12360440 ns         58   14.8135GB/s   632.042M items/s
BM_DatasetView_mixed_dimension_addition/16384/real_time                           24038376 ns   24038042 ns         30   15.2344GB/s   650.002M items/s
BM_DatasetView_mixed_dimension_addition/32768/real_time                           46886256 ns   46884652 ns         15   15.6212GB/s   666.507M items/s
BM_DatasetView_mixed_dimension_addition/65536/real_time                           92343760 ns   92342005 ns          7   15.8629GB/s   676.819M items/s
BM_DatasetView_mixed_dimension_addition/131072/real_time                         190677252 ns  190671359 ns          4   15.3646GB/s   655.558M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8/1/real_time                       32267 ns      32267 ns      19513   5.54169GB/s   236.445M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16/1/real_time                      61246 ns      61245 ns      11621   5.83923GB/s   249.141M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32/1/real_time                     126602 ns     126601 ns       5868   5.64966GB/s   241.052M items/s
BM_DatasetView_mixed_dimension_addition_threaded/64/1/real_time                     236870 ns     236869 ns       2823   6.03922GB/s   257.673M items/s
BM_DatasetView_mixed_dimension_addition_threaded/128/1/real_time                    446935 ns     446934 ns       1529   6.40143GB/s   273.128M items/s
BM_DatasetView_mixed_dimension_addition_threaded/256/1/real_time                    946541 ns     946538 ns        664   6.04522GB/s   257.929M items/s
BM_DatasetView_mixed_dimension_addition_threaded/512/1/real_time                   1902490 ns    1902484 ns        352   6.01532GB/s   256.654M items/s
BM_DatasetView_mixed_dimension_addition_threaded/1024/1/real_time                  3815311 ns    3815298 ns        179   5.99903GB/s   255.959M items/s
BM_DatasetView_mixed_dimension_addition_threaded/2048/1/real_time                  7498196 ns    7498122 ns         96   6.10498GB/s   260.479M items/s
BM_DatasetView_mixed_dimension_addition_threaded/4096/1/real_time                 15806796 ns   15806723 ns         47   5.79199GB/s   247.125M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8192/1/real_time                 29067949 ns   29067103 ns         23   6.29922GB/s   268.767M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16384/1/real_time                59597284 ns   59596279 ns         13   6.14476GB/s   262.176M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32768/1/real_time               120458336 ns  120451846 ns          5   6.08029GB/s   259.426M items/s
BM_DatasetView_mixed_dimension_addition_threaded/65536/1/real_time               269146323 ns  269141510 ns          3   5.44256GB/s   232.216M items/s
BM_DatasetView_mixed_dimension_addition_threaded/131072/1/real_time              483208747 ns  483192429 ns          2   6.06299GB/s   258.687M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8/2/real_time                       18868 ns      18868 ns      31981   9.47688GB/s   404.347M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16/2/real_time                      35308 ns      35308 ns      19078   10.1287GB/s   432.157M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32/2/real_time                      71463 ns      71460 ns      10238   10.0087GB/s    427.04M items/s
BM_DatasetView_mixed_dimension_addition_threaded/64/2/real_time                     133688 ns     133681 ns       5991   10.7003GB/s   456.548M items/s
BM_DatasetView_mixed_dimension_addition_threaded/128/2/real_time                    261554 ns     261550 ns       2580   10.9385GB/s   466.711M items/s
BM_DatasetView_mixed_dimension_addition_threaded/256/2/real_time                    489072 ns     489044 ns       1362   11.6998GB/s   499.192M items/s
BM_DatasetView_mixed_dimension_addition_threaded/512/2/real_time                   1004046 ns    1004022 ns        709    11.398GB/s   486.314M items/s
BM_DatasetView_mixed_dimension_addition_threaded/1024/2/real_time                  1983254 ns    1980872 ns        350   11.5407GB/s   492.404M items/s
BM_DatasetView_mixed_dimension_addition_threaded/2048/2/real_time                  4012300 ns    3990468 ns        181    11.409GB/s   486.784M items/s
BM_DatasetView_mixed_dimension_addition_threaded/4096/2/real_time                  7677860 ns    7677826 ns         88   11.9243GB/s   508.768M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8192/2/real_time                 15158091 ns   15086225 ns         41   12.0797GB/s   515.401M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16384/2/real_time                33472834 ns   31914873 ns         23   10.9405GB/s   466.796M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32768/2/real_time                71185300 ns   66363754 ns          9   10.2889GB/s   438.995M items/s
BM_DatasetView_mixed_dimension_addition_threaded/65536/2/real_time               135256625 ns  123844461 ns          5   10.8301GB/s   462.085M items/s
BM_DatasetView_mixed_dimension_addition_threaded/131072/2/real_time              257672188 ns  257658472 ns          2   11.3698GB/s   485.113M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8/4/real_time                       14151 ns      14151 ns      46251    12.636GB/s   539.137M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16/4/real_time                      23239 ns      23239 ns      30301   15.3889GB/s   656.594M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32/4/real_time                      39022 ns      39022 ns      18190   18.3296GB/s   782.062M items/s
BM_DatasetView_mixed_dimension_addition_threaded/64/4/real_time                      73798 ns      73797 ns       9295   19.3843GB/s   827.062M items/s
BM_DatasetView_mixed_dimension_addition_threaded/128/4/real_time                    147144 ns     147140 ns       4905   19.4437GB/s   829.598M items/s
BM_DatasetView_mixed_dimension_addition_threaded/256/4/real_time                    273867 ns     273862 ns       2453   20.8935GB/s   891.458M items/s
BM_DatasetView_mixed_dimension_addition_threaded/512/4/real_time                    555159 ns     555157 ns       1104   20.6141GB/s   879.535M items/s
BM_DatasetView_mixed_dimension_addition_threaded/1024/4/real_time                  1097378 ns    1097303 ns        692   20.8571GB/s   889.905M items/s
BM_DatasetView_mixed_dimension_addition_threaded/2048/4/real_time                  2124670 ns    2114495 ns        310   21.5452GB/s    919.26M items/s
BM_DatasetView_mixed_dimension_addition_threaded/4096/4/real_time                  5004344 ns    4930458 ns        157   18.2947GB/s   780.572M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8192/4/real_time                  8924192 ns    8294376 ns         64   20.5179GB/s   875.429M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16384/4/real_time                21174208 ns   18359172 ns         31   17.2951GB/s   737.926M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32768/4/real_time                44929395 ns   38108779 ns         17   16.3016GB/s   695.536M items/s
BM_DatasetView_mixed_dimension_addition_threaded/65536/4/real_time                95728657 ns   76845677 ns          7    15.302GB/s   652.887M items/s
BM_DatasetView_mixed_dimension_addition_threaded/131072/4/real_time              176855064 ns  149877745 ns          4   16.5655GB/s   706.793M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8/8/real_time                       11386 ns      11386 ns      62684   15.7044GB/s   670.055M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16/8/real_time                      13614 ns      13614 ns      46418   26.2688GB/s    1120.8M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32/8/real_time                      23383 ns      23383 ns      31082   30.5889GB/s   1.27454G items/s
BM_DatasetView_mixed_dimension_addition_threaded/64/8/real_time                      51062 ns      51010 ns      14190   28.0152GB/s    1.1673G items/s
BM_DatasetView_mixed_dimension_addition_threaded/128/8/real_time                    133595 ns     133590 ns       4353   21.4157GB/s   913.737M items/s
BM_DatasetView_mixed_dimension_addition_threaded/256/8/real_time                    179490 ns     179489 ns       3955   31.8795GB/s   1.32831G items/s
BM_DatasetView_mixed_dimension_addition_threaded/512/8/real_time                    340991 ns     340934 ns       2587   33.5613GB/s   1.39839G items/s
BM_DatasetView_mixed_dimension_addition_threaded/1024/8/real_time                   531606 ns     531604 ns       1022   43.0548GB/s   1.79395G items/s
BM_DatasetView_mixed_dimension_addition_threaded/2048/8/real_time                  1042457 ns    1041165 ns        670    43.912GB/s   1.82967G items/s
BM_DatasetView_mixed_dimension_addition_threaded/4096/8/real_time                  3024063 ns    2968683 ns        239   30.2747GB/s   1.26145G items/s
BM_DatasetView_mixed_dimension_addition_threaded/8192/8/real_time                  9786718 ns    6574040 ns         70   18.7096GB/s   798.276M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16384/8/real_time                22135136 ns   13503137 ns         30   16.5443GB/s   705.891M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32768/8/real_time                36343842 ns   23939527 ns         16   20.1526GB/s   859.843M items/s
BM_DatasetView_mixed_dimension_addition_threaded/65536/8/real_time                68396731 ns   39886346 ns         10   21.4169GB/s   913.786M items/s
BM_DatasetView_mixed_dimension_addition_threaded/131072/8/real_time              158752977 ns   86635223 ns          5   18.4544GB/s   787.387M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8/16/real_time                      11609 ns      11609 ns      57873   15.4033GB/s   657.209M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16/16/real_time                     15574 ns      15566 ns      44600   22.9634GB/s   979.771M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32/16/real_time                     24893 ns      24886 ns      28769   28.7329GB/s    1.1972G items/s
BM_DatasetView_mixed_dimension_addition_threaded/64/16/real_time                     42421 ns      42416 ns      16597   33.7215GB/s   1.40506G items/s
BM_DatasetView_mixed_dimension_addition_threaded/128/16/real_time                    78586 ns      78586 ns       8888   36.4063GB/s   1.51693G items/s
BM_DatasetView_mixed_dimension_addition_threaded/256/16/real_time                   150706 ns     150563 ns       4666   37.9683GB/s   1.58201G items/s
BM_DatasetView_mixed_dimension_addition_threaded/512/16/real_time                   291547 ns     291399 ns       2375    39.253GB/s   1.63554G items/s
BM_DatasetView_mixed_dimension_addition_threaded/1024/16/real_time                  585242 ns     584132 ns       1210   39.1089GB/s   1.62954G items/s
BM_DatasetView_mixed_dimension_addition_threaded/2048/16/real_time                 1183896 ns    1171434 ns        603   38.6659GB/s   1.61108G items/s
BM_DatasetView_mixed_dimension_addition_threaded/4096/16/real_time                 2757646 ns    2757599 ns        243   33.1996GB/s   1.38332G items/s
BM_DatasetView_mixed_dimension_addition_threaded/8192/16/real_time                10500788 ns    7110079 ns         69   17.4373GB/s   743.992M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16384/16/real_time               22026728 ns   17820405 ns         33   16.6258GB/s   709.365M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32768/16/real_time               36567654 ns   22010720 ns         18   20.0292GB/s    854.58M items/s
BM_DatasetView_mixed_dimension_addition_threaded/65536/16/real_time               86694637 ns   36988220 ns          8   16.8966GB/s   720.921M items/s
BM_DatasetView_mixed_dimension_addition_threaded/131072/16/real_time             139327624 ns   76831023 ns          7   21.0273GB/s   897.166M items/s
BM_DatasetView_mixed_dimension_addition_threaded/8/24/real_time                      25025 ns      23929 ns      54176   7.14544GB/s   304.872M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16/24/real_time                     17340 ns      17230 ns      45934    20.624GB/s   879.957M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32/24/real_time                     47802 ns      44891 ns      31948   14.9629GB/s   638.419M items/s
BM_DatasetView_mixed_dimension_addition_threaded/64/24/real_time                     34302 ns      34195 ns      16936    41.704GB/s   1.73767G items/s
BM_DatasetView_mixed_dimension_addition_threaded/128/24/real_time                   131971 ns     124901 ns       9772   21.6791GB/s   924.976M items/s
BM_DatasetView_mixed_dimension_addition_threaded/256/24/real_time                   115861 ns     115441 ns       6449   49.3874GB/s   2.05781G items/s
BM_DatasetView_mixed_dimension_addition_threaded/512/24/real_time                   520206 ns     483366 ns       1991   21.9992GB/s   938.631M items/s
BM_DatasetView_mixed_dimension_addition_threaded/1024/24/real_time                  412636 ns     408465 ns       1523   55.4683GB/s   2.31118G items/s
BM_DatasetView_mixed_dimension_addition_threaded/2048/24/real_time                 1695689 ns    1553307 ns        625   26.9957GB/s   1.12482G items/s
BM_DatasetView_mixed_dimension_addition_threaded/4096/24/real_time                 3025353 ns    2866432 ns        286   30.2618GB/s   1.26091G items/s
BM_DatasetView_mixed_dimension_addition_threaded/8192/24/real_time                11872702 ns    8885767 ns         64   15.4224GB/s   658.022M items/s
BM_DatasetView_mixed_dimension_addition_threaded/16384/24/real_time               20723976 ns   11384367 ns         35   17.6709GB/s   753.958M items/s
BM_DatasetView_mixed_dimension_addition_threaded/32768/24/real_time               41010408 ns   18984886 ns         17   17.8594GB/s   762.002M items/s
BM_DatasetView_mixed_dimension_addition_threaded/65536/24/real_time               80677987 ns   29473699 ns          8   18.1567GB/s   774.685M items/s
BM_DatasetView_mixed_dimension_addition_threaded/131072/24/real_time             160858489 ns   51066397 ns          4   18.2128GB/s   777.081M items/s
BM_DatasetView_multi_column_mixed_dimension_nested/8                                 12860 ns      12860 ns      49378    13.905GB/s   607.521k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/16                                24672 ns      24669 ns      25746   14.4973GB/s   633.398k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/32                                48031 ns      48030 ns      14396   14.8918GB/s   650.633k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/64                                94813 ns      94809 ns       6931   15.0884GB/s    659.22k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/128                              190316 ns     190304 ns       3906    15.034GB/s   656.844k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/256                              388850 ns     388822 ns       1824   14.7164GB/s   642.968k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/512                              796478 ns     796388 ns        874     14.37GB/s   627.835k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/1024                            1658560 ns    1658518 ns        421   13.8004GB/s   602.948k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/2048                            3489555 ns    3489347 ns        201   13.1189GB/s   573.173k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/4096                            6921331 ns    6920970 ns        104   13.2283GB/s   577.954k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/8192                           14698912 ns   14698914 ns         38   12.4571GB/s   544.258k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/16384                          29940487 ns   29936687 ns         23   12.2328GB/s   534.461k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/32768                          61735700 ns   61735522 ns         11   11.8639GB/s    518.34k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/65536                         145125393 ns  145122188 ns          6   10.0939GB/s   441.008k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/131072                        263633529 ns  263629653 ns          3   11.1129GB/s    485.53k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/262144                        515697222 ns  515681806 ns          2   11.3624GB/s    496.43k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8/1/real_time            13226 ns      13226 ns      54270   13.5201GB/s   590.701k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16/1/real_time           25320 ns      25320 ns      27863   14.1245GB/s   617.109k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32/1/real_time           50421 ns      50420 ns      12960   14.1858GB/s   619.788k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/64/1/real_time           96990 ns      96988 ns       6933   14.7491GB/s   644.399k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/128/1/real_time         195751 ns     195748 ns       3658   14.6157GB/s   638.568k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/256/1/real_time         394838 ns     394827 ns       1709   14.4922GB/s   633.172k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/512/1/real_time         732952 ns     732927 ns        863   15.6137GB/s   682.173k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/1024/1/real_time       1590315 ns    1590285 ns        438   14.3922GB/s   628.806k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/2048/1/real_time       3418216 ns    3418144 ns        194   13.3919GB/s   585.101k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/4096/1/real_time       7086847 ns    7086639 ns        105   12.9187GB/s   564.426k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8192/1/real_time      15003295 ns   15002970 ns         48   12.2044GB/s   533.216k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16384/1/real_time     29019226 ns   29018159 ns         25   12.6196GB/s   551.359k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32768/1/real_time     59039069 ns   59038835 ns         12   12.4057GB/s   542.014k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/65536/1/real_time    134860121 ns  134859561 ns          6   10.8619GB/s   474.566k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/131072/1/real_time   255925653 ns  255919517 ns          3   11.4474GB/s   500.145k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/262144/1/real_time   511710451 ns  511707645 ns          1   11.4506GB/s   500.283k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8/2/real_time            10352 ns      10352 ns      59112   17.2733GB/s   754.682k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16/2/real_time           18219 ns      18219 ns      41815   19.6289GB/s   857.601k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32/2/real_time           28948 ns      28948 ns      23577    24.708GB/s   1079.51k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/64/2/real_time           54061 ns      54061 ns      12503   26.4609GB/s     1.129M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/128/2/real_time         105832 ns     105832 ns       6433   27.0336GB/s   1.15343M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/256/2/real_time         203650 ns     203644 ns       3621   28.0975GB/s   1.19883M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/512/2/real_time         416689 ns     416678 ns       1797   27.4643GB/s   1.17181M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/1024/2/real_time        865264 ns     865246 ns        822   26.4522GB/s   1.12863M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/2048/2/real_time       1988958 ns    1988947 ns        360   23.0153GB/s   1005.55k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/4096/2/real_time       8357291 ns    6081662 ns         83   10.9548GB/s   478.624k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8192/2/real_time      18292107 ns   11891846 ns         38   10.0101GB/s   437.347k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16384/2/real_time     37882670 ns   23742298 ns         19   9.66698GB/s   422.357k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32768/2/real_time     72821568 ns   36901248 ns          8   10.0578GB/s    439.43k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/65536/2/real_time    140883475 ns   70261567 ns          5   10.3976GB/s   454.276k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/131072/2/real_time   287791045 ns  125353690 ns          2   10.1799GB/s   444.767k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/262144/2/real_time   566379181 ns  248151623 ns          1   10.3453GB/s   451.994k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8/4/real_time             7976 ns       7976 ns      77988   22.4186GB/s   979.482k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16/4/real_time           10882 ns      10882 ns      66634   32.8637GB/s   1.40218M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32/4/real_time           17288 ns      17287 ns      40601   41.3741GB/s   1.76529M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/64/4/real_time           31349 ns      31349 ns      22339   45.6317GB/s   1.94695M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/128/4/real_time          57460 ns      57460 ns      12007   49.7917GB/s   2.12445M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/256/4/real_time         110822 ns     110819 ns       6894   51.6329GB/s     2.203M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/512/4/real_time         219415 ns     219414 ns       3215   52.1573GB/s   2.22538M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/1024/4/real_time        421315 ns     419847 ns       1787   54.3255GB/s   2.31789M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/2048/4/real_time       1357031 ns    1356999 ns        435   33.7327GB/s   1.43926M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/4096/4/real_time       8254135 ns    4608686 ns         84   11.0917GB/s   484.606k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8192/4/real_time      20714569 ns   12321131 ns         67   8.83945GB/s   386.202k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16384/4/real_time     23288937 ns   16676029 ns         29   15.7247GB/s   687.021k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32768/4/real_time     41932922 ns   29760332 ns         16   17.4665GB/s   763.124k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/65536/4/real_time     88459302 ns   56230006 ns          8   16.5595GB/s   723.497k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/131072/4/real_time   168789639 ns  118711871 ns          4    17.357GB/s    758.34k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/262144/4/real_time   389263398 ns  193923354 ns          2   15.0525GB/s   657.652k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8/8/real_time             7548 ns       7547 ns      83329   23.6918GB/s   1035.11k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16/8/real_time            8888 ns       8888 ns      75487   40.2381GB/s   1.71683M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32/8/real_time           12117 ns      12117 ns      55658   59.0299GB/s   2.51861M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/64/8/real_time           19897 ns      19897 ns      37038   71.8951GB/s   3.06753M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/128/8/real_time          32897 ns      32896 ns      21319   86.9703GB/s   3.71073M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/256/8/real_time          60006 ns      60005 ns      11391   95.3586GB/s   4.06863M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/512/8/real_time         125532 ns     125529 ns       5739   91.1649GB/s    3.8897M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/1024/8/real_time        217421 ns     217414 ns       3271   105.271GB/s   4.49157M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/2048/8/real_time       1553084 ns    1548261 ns        464   29.4745GB/s   1.25758M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/4096/8/real_time       6411999 ns    4277892 ns         83   14.2783GB/s    623.83k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8192/8/real_time      10426394 ns    7160725 ns         61   17.5617GB/s   767.283k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16384/8/real_time     26980973 ns   13413331 ns         24   13.5729GB/s    593.01k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32768/8/real_time     33659774 ns   29299326 ns         16   21.7596GB/s    950.69k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/65536/8/real_time     75597757 ns   57395497 ns         10   19.3768GB/s   846.586k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/131072/8/real_time   289684862 ns   66465661 ns          4   10.1134GB/s   441.859k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/262144/8/real_time   550057107 ns  135444407 ns          1   10.6523GB/s   465.406k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8/16/real_time           11127 ns      11127 ns      66218   16.0705GB/s   702.129k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16/16/real_time          11683 ns      11683 ns      59816   30.6106GB/s   1.30605M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32/16/real_time          15273 ns      15269 ns      45079   46.8325GB/s   1.99819M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/64/16/real_time          21512 ns      21511 ns      32471   66.4993GB/s    2.8373M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/128/16/real_time         34704 ns      34694 ns      19773   82.4413GB/s    3.5175M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/256/16/real_time         61743 ns      61742 ns      11291   92.6759GB/s   3.95417M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/512/16/real_time        114024 ns     114022 ns       6145   100.366GB/s   4.28227M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/1024/16/real_time       218634 ns     218633 ns       3192   104.687GB/s   4.46666M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/2048/16/real_time       874264 ns     874111 ns        695   52.3599GB/s   2.23402M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/4096/16/real_time      7491123 ns    4671132 ns         92   12.2215GB/s   533.965k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8192/16/real_time     15446151 ns    7010782 ns         43   11.8544GB/s   517.928k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16384/16/real_time    28598352 ns   13477990 ns         46   12.8053GB/s   559.473k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32768/16/real_time    58651543 ns   23489319 ns         11   12.4877GB/s   545.595k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/65536/16/real_time   123541434 ns   33525543 ns          5   11.8571GB/s   518.045k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/131072/16/real_time  208780399 ns   89140219 ns          3   14.0324GB/s   613.084k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/262144/16/real_time  472445841 ns  162656303 ns          2   12.4022GB/s   541.861k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8/24/real_time           14508 ns      14391 ns      41122   12.3249GB/s   538.483k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16/24/real_time          16109 ns      16031 ns      41595    22.201GB/s   969.976k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32/24/real_time          80845 ns      73900 ns      10000   8.84724GB/s   386.542k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/64/24/real_time          43007 ns      40707 ns      18382   33.2619GB/s   1.41918M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/128/24/real_time         70931 ns      63808 ns      20558   40.3355GB/s   1.72098M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/256/24/real_time         65938 ns      63108 ns      13607   86.7787GB/s   3.70256M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/512/24/real_time        742749 ns     655880 ns       1000   15.4077GB/s   673.175k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/1024/24/real_time       164775 ns     163180 ns       4181   138.906GB/s   5.92666M items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/2048/24/real_time      1925015 ns    1859985 ns        417   23.7797GB/s   1038.95k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/4096/24/real_time      7185131 ns    5894474 ns         73    12.742GB/s   556.705k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/8192/24/real_time     17895510 ns    8355631 ns        101   10.2319GB/s    447.04k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/16384/24/real_time    34499935 ns   15859307 ns         20   10.6148GB/s   463.769k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/32768/24/real_time    69211120 ns   20649167 ns          9   10.5824GB/s   462.353k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/65536/24/real_time   125431134 ns   30042559 ns          5   11.6785GB/s    510.24k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/131072/24/real_time  226112134 ns   70664221 ns          3   12.9568GB/s   566.091k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_threaded/262144/24/real_time  515139353 ns  138725286 ns          2   11.3743GB/s   496.953k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/8                       25103 ns      25102 ns      29721   7.12337GB/s   311.225k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/16                      56105 ns      56104 ns      12726   6.37434GB/s   278.499k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/32                     148791 ns     148787 ns       4624   4.80724GB/s   210.032k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/64                     294330 ns     294329 ns       2283   4.86024GB/s   212.347k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/128                    590205 ns     590203 ns       1232   4.84752GB/s   211.791k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/256                   1355832 ns    1355828 ns        568   4.22033GB/s   184.389k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/512                   8855208 ns    8854957 ns         86   1.29239GB/s   56.4655k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/1024                 20170568 ns   20170103 ns         36   1.13476GB/s   49.5783k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/2048                 40985846 ns   40985645 ns         16   1.11689GB/s   48.7976k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/4096                 94878777 ns   94876681 ns          8   988.125MB/s     42.16k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/8192                271548041 ns  271545290 ns          3   690.493MB/s    29.461k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/16384               671099652 ns  671086897 ns          1   558.795MB/s   23.8419k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/32768              1388271574 ns 1388235758 ns          1   540.254MB/s   23.0508k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/65536              2932553500 ns 2932484089 ns          1   511.512MB/s   21.8245k items/s
```